### DISCUSSION TITLE
auto loading preset with wildcard

### DISCUSSION THREAD

**Discussion by funny0facer** [wwmm/easyeffects#3157] [Q&A]
Is there a way to use wildcards when autoloading presets? Everytime, I log out and in again, the number of **alsa_output.pci-0000_12_00.6.analog-stereo.2** increases by 1.

I tried to add a wildcard (*) at the end of the name, but it was not successful
![grafik](https://github.com/wwmm/easyeffects/assets/75338570/e41584e5-c2f6-4f97-8150-9861480bc59d)

Is there a way to use a wildcard in the device name?

thanks!

**Comment by wwmm** [comment]
Honestly I am not sure if it is doable. We do check if the name contains the bus id from the device that went through routing changes https://github.com/wwmm/easyeffects/blob/b26f456cfc4f8035d2c513cbbee842ab9cd03f4e/src/application.cpp#L162C100-L162C106. But I am afraid that in a system with 2 or more soundcards a wildcard approach may get the wrong device.

As PipeWire is right now the autoloading is a little tricky to do. We have to listen to routing changes in device objects. But what the user actually selects are what are called node objects. When the routing changes we have search through the node list and finding which is the corresponding device.

**Reply by wwmm** [reply]
The problem with a wildcard is that we may try to load a preset for routing changes that happend to devices for which the user did not want an autoloading. The ideal solution would be to understand why your bus id number are changes so often. They unfortunately do change eventually for reasons I never understood. It happens on my computer. But not every time I log out.

**Reply by violetmage** [reply]
I'm very curious about why the number goes up.  Perhaps it's funky drivers again.

@funny0facer What happens when you reboot your computer? Does the number go back to where it started?

**Reply by funny0facer** [reply]
after a reboot, the number starts at 2 again. So no big deal as it is a very uncommen workflow to log off and in so often.
My workaround is now to just have 10 times the autoload config file there with different numbers.

if it is important, I use KDE

**Reply by violetmage** [reply]
Interesting. @funny0facer Thanks for the info.

@wwmm I'm starting to wonder if this has something to do with wireplumber

**Reply by funny0facer** [reply]
I can confirm. I just switched back to pipewire-media-session and the behavior is gone.
it is just "alsa_output.pci-0000_12_00.6.analog-stereo" now.

**Reply by violetmage** [reply]
@funny0facer Thanks for the help

Will definitely have to do some more investigating when I have time.

**Reply by wwmm** [reply]
> I can confirm. I just switched back to pipewire-media-session and the behavior is gone. it is just "alsa_output.pci-0000_12_00.6.analog-stereo" now.

I would never have suspected wireplumber. If installing the legacy session manager fixes it it is worth to open an issue at wireplumber's page asking about if this is the expected behavior or a bug.

**Reply by funny0facer** [reply]
created a ticket there. If you want to follow: https://gitlab.freedesktop.org/pipewire/wireplumber/-/issues/669

**Reply by wwmm** [reply]
> created a ticket there. If you want to follow: https://gitlab.freedesktop.org/pipewire/wireplumber/-/issues/669

Thanks! I've enabled notifications for messages there so I can see what they say.

**Comment by funny0facer** [comment]
Nevermind, I found a good enough workaround by renaming the node.name by myself with wireplumber rules.

```
monitor.alsa.rules = [
  {
    matches = [
      {
        device.name = "alsa_card.pci-0000_12_00.6"
      }
    ]
    actions = {
      update-props = {
        device.description = "Motherboard"
      }
    }
  },
  {
    matches = [
      {
        node.description = "Motherboard Analog Stereo"
      }
    ]
    actions = {
      update-props = {
        node.name = "alsa_output.pci-0000_12_00.6.analog-stereo"
      }
    }
  }
]
```
